infrastructure/logger/level.go

```diff
@@ -0,0 +1,73 @@
+package logger
+
+import "strings"
+
+// Level is the level at which a logger is configured. All messages sent
+// to a level which is below the current level are filtered.
+type Level uint32
+
+// Level constants.
+const (
+	LevelTrace Level = iota
+	LevelDebug
+	LevelInfo
+	LevelWarn
+	LevelError
+	LevelCritical
+	LevelOff
+)
+
+// levelStrs defines the human-readable names for each logging level.
+var levelStrs = [...]string{"TRC", "DBG", "INF", "WRN", "ERR", "CRT", "OFF"}
+
+// LevelFromString returns a level based on the input string s. If the input
+// can't be interpreted as a valid log level, the info level and false is
+// returned.
+func LevelFromString(s string) (l Level, ok bool) {
+	switch strings.ToLower(s) {
+	case "trace", "trc":
+		return LevelTrace, true
+	case "debug", "dbg":
+		return LevelDebug, true
+	case "info", "inf":
+		return LevelInfo, true
+	case "warn", "wrn":
+		return LevelWarn, true
+	case "error", "err":
+		return LevelError, true
+	case "critical", "crt":
+		return LevelCritical, true
+	case "off":
+		return LevelOff, true
+	default:
+		return LevelInfo, false
+	}
+}
+
+// String returns the tag of the logger used in log messages, or "OFF" if
+// the level will not produce any log output.
+func (l Level) String() string {
+	if l >= LevelOff {
+		return "OFF"
+	}
+	return levelStrs[l]
+}
+
+// validLogLevel returns whether or not logLevel is a valid debug log level.
+func validLogLevel(logLevel string) bool {
```

Comment:
It seems like this function is redundant, since it's always called a few moments before SetLogLevel is called, which re-validates the log level string.

However, if you think it's still needed, why not `_, ok := LevelFromString; return ok`?
This will re-use the hard-coded strings in LevelFromString.

Alternatively, create a map between strings to levels, and then use then in both functions.

Or even better - enjoy both worlds - create a map between string to levels just to avoid hard-coded strings, and then delete validLogLevel/re-use LevelFromString because reasons from above.

Reply:
But with a map I need to care about concurrency (or assume no one will ever modify that map)